### PR TITLE
Adapt `NetworkPolicy`s for components in `Istio`-related namespaces

### PR DESCRIPTION
**Pull request by timuthy**
**How to categorize this PR?**

/area dev-productivity ops-productivity networking security
/kind enhancement

**What this PR does / why we need it**:
This PR adapts and replaces all possible network policies that are used for `Istio-{System, Ingress}` with the help of the [NetworkPolicy](https://github.com/gardener/gardener/tree/master/pkg/resourcemanager/controller/networkpolicy) controller.

Co-authored-by: Rafael Franzke <rafael.franzke@sap.com>

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/7352

**Special notes for your reviewer**:
f10d67bf8ddd0277a03524ceb2674af2ec0d788c adds required labels for exposureclass handler namespaces along the way(cc @ScheererJ @axel7born). This was probably missed before.

**Release note**:

```other developer
Network policies for `Istio-System` and `Istio-Ingress` have been adapted and now make use of GRM's [NetworkPolicy Controller](https://github.com/gardener/gardener/blob/master/docs/concepts/resource-manager.md#networkpolicy-controller).
```

### CODE REVIEW

File: pkg/operation/botanist/component/resourcemanager/resource_manager.go
Comment (rfranzke) [inline pkg/operation/botanist/component/resourcemanager/resource_manager.go]: Don't we need https://github.com/gardener/gardener/blob/278a36cd978dce8f3ea1c3f2a946c9afbd55b93e/pkg/gardenlet/controller/networkpolicy/reconciler.go#L210 as well? cc @axel7born

File: pkg/operation/botanist/component/istio/charts/istio/istio-istiod/templates/service.yaml
Comment (rfranzke) [inline pkg/operation/botanist/component/istio/charts/istio/istio-istiod/templates/service.yaml]:
```suggestion
    networking.resources.gardener.cloud/from-world-to-ports: '[{"port":{{ .Values.ports.https }},"protocol":"TCP"}]'
```

This should not be needed, or?

File: pkg/operation/botanist/component/vpnshoot/vpnshoot.go
Comment (rfranzke) [inline pkg/operation/botanist/component/vpnshoot/vpnshoot.go]: Can you add a comment here why this is needed?

File: cmd/gardenlet/app/app.go
Comment (rfranzke) [inline cmd/gardenlet/app/app.go]:
```suggestion
	for _, serviceName := range []string{vpnseedserver.ServiceName, vpnseedserver.ServiceName+"-0",vpnseedserver.ServiceName+"-1"} {
```

File: pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go
Comment (rfranzke) [inline pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go:74]: Why is this public now?

File: cmd/gardenlet/app/app.go
Comment (rfranzke) [inline cmd/gardenlet/app/app.go:503]: Shouldn't it be like this?

```suggestion
		taskFns = append(taskFns, migrationTasksForServices(g.mgr.GetClient(), serviceList.Items, vpnseedserver.OpenVPNPort)...)
```

File: pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_service.go
Comment (rfranzke) [inline pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_service.go]: Instead of configuring this explicitly here, can't we check it implicitly by checking whether the service's namespace starts with https://github.com/gardener/gardener/blob/3151ac7a5c086a9ee7a8707bf6cc33f4ba0a5988/pkg/apis/core/v1beta1/constants/types_constants.go#L797-L799?

**Review by rfranzke — CHANGES_REQUESTED** [review]
Can you now also make the `NetworkPolicy` controller create the `deny-all` policy in the namespaces? https://github.com/gardener/gardener/blob/278a36cd978dce8f3ea1c3f2a946c9afbd55b93e/pkg/gardenlet/controller/networkpolicy/reconciler.go#L140-L147

/cc @ScheererJ @MartinWeindel 
Please have a look as well.

File: pkg/operation/botanist/component/resourcemanager/resource_manager.go
Comment (ScheererJ) [reply pkg/operation/botanist/component/resourcemanager/resource_manager.go]: Yes, indeed. We need this for the exposure class istio namespaces as well. Unfortunately, it is not that straight forward as `v1beta1constants.GardenRoleIstioSystem` is the prefix of all potential values (see https://github.com/gardener/gardener/blob/6764bf9a4953c26331a2872e903a2ebaba112603/pkg/operation/istio_config.go#L128).
In hindsight, it might not have been a good idea to append the zone to the garden role value.

Not sure how to solve this in a good way.

File: pkg/operation/botanist/component/istio/charts/istio/istio-istiod/templates/service.yaml
Comment (ScheererJ) [inline pkg/operation/botanist/component/istio/charts/istio/istio-istiod/templates/service.yaml]: Similarly, this is not sufficient if exposure classes are used with multiple availability zones. Then, the gardener role value will be something like `exposureclass-handler--zone--eu-west-1c`.

File: pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_service.go
Comment (ScheererJ) [inline pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_service.go]: Again, with zonal exposure classes this is not sufficient, unfortunately.

File: pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_service_test.go
Comment (ScheererJ) [inline pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_service_test.go]: Once again, zonal exposure classes need to be handled differently here.

File: pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_service.go
Comment (ScheererJ) [inline pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_service.go]: Unfortunately, zonal exposure classes require a different handling here.

File: pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_service.go
Comment (ScheererJ) [inline pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_service.go:158]: I am a bit confused. Why does this not include the exposure classes at all (see line 172)?

File: pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go
Comment (ScheererJ) [inline pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go]: Zonal exposure classes need to be handled as well.

**Review by ScheererJ — CHANGES_REQUESTED** [review]
Zonal exposure classes are currently not handled properly. Please adapt.

File: pkg/operation/botanist/component/istio/charts/istio/istio-istiod/templates/service.yaml
Comment (rfranzke) [reply pkg/operation/botanist/component/istio/charts/istio/istio-istiod/templates/service.yaml]: Well, this we have to fix anyways. We need a static label for all namespaces running istio-ingress components. @axel7born Wanted to look into also using `gardener.cloud/role=istio-ingress` for all namespaces related to exposure classes, but it isn't done yet.

File: pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_service.go
Comment (timuthy) [reply pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_service.go]: Yes, good idea. I will change it.

File: pkg/operation/botanist/component/istio/charts/istio/istio-istiod/templates/service.yaml
Comment (timuthy) [reply pkg/operation/botanist/component/istio/charts/istio/istio-istiod/templates/service.yaml]: I see this as a prerequisite to drop the already existing network policies.
What do you think about not dropping the existing policies, but first apply the common label and then remove the said policies in a later release?

Comment (ScheererJ) [reply pkg/operation/botanist/component/istio/charts/istio/istio-istiod/templates/service.yaml]: Fine with me.

File: cmd/gardenlet/app/app.go
Comment (timuthy) [reply cmd/gardenlet/app/app.go:503]: The port is used for the `all-scrape-targets` annotation. Why should we need the OpenVPN Port here?

File: pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go
Comment (timuthy) [reply pkg/operation/botanist/component/vpnseedserver/vpn_seed_server.go:74]: It's used for the one-time migration in `gardenlet/app/app.go`.

File: pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_service.go
Comment (timuthy) [reply pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_service.go]: Thinking about it again, this will be problematic for CP namespaces created a long time ago when we used the scheme `shoot-<project-name>-<shoot-name>`. To be really sure, we can use the label selector with `gardener.cloud/role: shoot`.
WDYT?

File: cmd/gardenlet/app/app.go
Comment (rfranzke) [reply cmd/gardenlet/app/app.go:503]: Right. My bad. Ignore this

File: pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_service.go
Comment (rfranzke) [reply pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_service.go]: Why would it be a problem? The approach with checking for `TechnicalIDPrefix` also works well for namespaces with the "old scheme"?

Comment (timuthy) [reply pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_service.go]: My bad, I expected the technical prefix to be `shoot--`. Thanks 👍

File: pkg/operation/botanist/component/istio/charts/istio/istio-istiod/templates/service.yaml
Comment (timuthy) [reply pkg/operation/botanist/component/istio/charts/istio/istio-istiod/templates/service.yaml:8]: I added a special handling for zonal exposure classes in e747c53164eba9966fd71c4ce692c5f3665165b9. I hope that's fine.

File: pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_service.go
Comment (timuthy) [reply pkg/operation/botanist/component/kubeapiserverexposure/kube_apiserver_service.go:158]: The exposure class label selector is only added if it's a shoot namespace. I assume that when the service is applied by the Gardener-Operator, we (hopefully) don't need any exposure class handling (yet) 😅

File: pkg/operation/botanist/component/resourcemanager/resource_manager.go
Comment (timuthy) [reply pkg/operation/botanist/component/resourcemanager/resource_manager.go:567]: See https://github.com/gardener/gardener/pull/7570#discussion_r1120169974.

**Comment by timuthy** [comment]
Most of the requested changes were pushed in separate commits. The PR still missed the `deny-all` handling (https://github.com/gardener/gardener/pull/7570#pullrequestreview-1315000986) which will follow in a bit.

File: pkg/operation/botanist/component/vpnshoot/vpnshoot.go
Comment (rfranzke) [inline pkg/operation/botanist/component/vpnshoot/vpnshoot.go]: not needed since we already have https://github.com/gardener/gardener/pull/7570/files#diff-d209ad66b99a63c3b888ee941c5fc40f2fd54823e01bc353045551af0a6850a2R7

**Comment by timuthy** [comment]
/hold
Together with @rfranzke we discussed that https://github.com/gardener/gardener/commit/e747c53164eba9966fd71c4ce692c5f3665165b9 can be reverted and GRM can be activated for zonal exposure class by using the label value `handler.exposureclass.gardener.cloud/name` (cc @ScheererJ)

File: pkg/operation/botanist/component/vpnshoot/vpnshoot.go
Comment (timuthy) [reply pkg/operation/botanist/component/vpnshoot/vpnshoot.go]: Tackled in 37f5002d42f1b0f204cf9f074922f5bd493a7bba.

File: pkg/operation/botanist/component/istio/charts/istio/istio-istiod/templates/service.yaml
Comment (timuthy) [reply pkg/operation/botanist/component/istio/charts/istio/istio-istiod/templates/service.yaml]: Tackled in 12d5f31f48066f50d836c08d09596a3b73622986.

**Comment by timuthy** [comment]
Thanks for your feedback @rfranzke and @ScheererJ. I addressed all comments, please see commits cb429b5bf8714563cba39659bb012b3ff946adc6 and following.

File: cmd/gardenlet/app/app.go
Comment (ScheererJ) [inline cmd/gardenlet/app/app.go]: Is it not required to change this as well, now that we use the exposure class name label in other places?

**Review by ScheererJ — CHANGES_REQUESTED** [review]
Looks mostly good now. Can you please check whether the migration needs to be adapted to the new exposure class label handling?

File: cmd/gardenlet/app/app.go
Comment (timuthy) [reply cmd/gardenlet/app/app.go]: Good catch, I changed it as well. Thank you 🙂

**Review by ScheererJ — APPROVED** [review]
/lgtm

**Comment by timuthy** [comment]
> New changes are detected. LGTM label has been removed.

Force pushed after rebase due to merge conflicts with https://github.com/gardener/gardener/commit/c35f375c4f6f17eaac32afd00edb9f0705ab4b8f.

**Review by rfranzke — COMMENTED** [review]
Awesome, thanks! Let's get this in.

/lgtm
/approve

**Comment by timuthy** [comment]
Thanks again for your input and support @ScheererJ and @rfranzke 👍

**Pull request merged**
